### ISSUE TITLE
Segmentation fault when encountering a TemplateHaskell splice

### ISSUE THREAD

**Issue by afcowie** [digital-asset/ghcide#444]
We've had **ghcide** crashing, and the editor trying to reload it leads to a tight loop that the editor can't escape. Trying the _ghcide_ binary from the command-line in test mode leads to a segfault:

```
Step 6/6: Type checking the files
Segmentation fault (core dumped)
```

This is triggered when using TemplateHaskell. We have a splice [`fromPackage :: Q Exp`](https://hackage.haskell.org/package/core-program-0.2.4.2/docs/Core-Program-Metadata.html#v:fromPackage) which returns some metadata from the project's _.cabal_ file:

```haskell
version :: Version
version = $(fromPackage)
```

removing this definition from the source file (in this case [_RenderMain.hs_](https://github.com/oprdyn/publish/blob/master/src/RenderMain.hs#L13-L14) lines 13-14) allows the _ghcide_ binary to complete its test run:

```
Step 6/6: Type checking the files

Completed (1 file worked, 0 files failed)
```
I had earlier speculated that QuasiQuotes were to blame, but the file checks ok with those `[quote| ... |]` blocks still in the file.

Other files in the project load and check fine, so long as you avoid having the problematic file open in an editor tab you can avoid the crash. Not ideal, but at least a workaround.

**Comment by ndmitchell** [comment]
Another workaround is to use `#ifdef __GHCIDE__` around the problematic bit.

**Comment by afcowie** [comment]
Fencing the splice with a preprocessor directive (and enabling `CPP`) indeed works to prevent _ghcide_ from crashing.

```haskell
version :: Version
#ifdef __GHCIDE__
version = "0"
#else
version = $(fromPackage)
#endif
```

Do you want a backtrace from when it segfaults? I've not had much success getting GDB to play nice with GHC but I'll give it a try if you want.

**Comment by ndmitchell** [comment]
If it's easy then sure, but I'm not sure if anyone is likely to look at it. TH related things aren't our strong spot.

**Comment by mpickering** [comment]
Is this reproducible @afcowie ?

**Comment by afcowie** [comment]
@mpickering yes, very reproducible. I only have one Template Haskell splice in my collection of projects—only time I've ever written TH code—but I hazard a guess any Template Haskell splice would trigger this.

**Comment by mpickering** [comment]
Template Haskell does generally work so I think it is probably something specific to do with this splice. 
What platform are you running on? What GHC version?

**Comment by mpickering** [comment]
`fromPackage` looks excessively dodgy. Perhaps the issue is that it is calling `error` which manifests as a segfault somehow.

**Comment by afcowie** [comment]
@mpickering I'm sorry you find our code excessively dodgy. We've avoided TemplateHaskell for years, seems to cause more problems that its worth but worked for us in this case. I'll try your suggestion.

**Comment by afcowie** [comment]
@mpickering No, changing to `fail` doesn't change the outcome; still segfaults. I'll poke around in there to see if there's something else going on though.

**Comment by mpickering** [comment]
@afcowie Can you produce a standalone repo with a `hie.yaml` file which reproduces this and I will investigate.

**Comment by afcowie** [comment]
@mpickering done. If you clone https://github.com/afcowie/hello/tree/ghcide-issue-444 you will get a essentially bare repository with a single program that demonstrates the problem. `stack build` should succeed; `stack exec -- hello` and/or `stack run` should result in "Hello World" on your console:

```
$ stack exec -- hello
Hello World
$
```

however
```
$ ghcide src/HelloMain.hs
Step 6/6: Type checking the files
Segmentation fault (core dumped)
$
```
I should note that this is with GHC 8.8.2 so you'll need the a version of _ghcide_ built against that, as I understand it. The problem was exhibited with earlier versions of GHC. If you're using Stack but GHC 8.6.5, say, you can use:

```yaml
resolver: lts-14.27

extra-deps:
  - core-text-0.2.3.3
  - core-data-0.2.1.5
  - core-program-0.2.4.2
```
and get the same demonstration.

**Comment by mpickering** [comment]
The segfault is coming from calling `length` in `Data.Text.Short`. Here is a more minimal example.

https://gist.github.com/d335ac940a97a574891c947578636fa1

**Comment by mpickering** [comment]
The `ByteArray#` and length look correct to me so something is going wrong with the call to `c_text_short_length`.

**Comment by mpickering** [comment]
To further deepen the mystery, if I add `text-short` as a local package, then it works, if it comes from the global cabal store then it doesn't. I'm going to stop investigating this bug now!

**Comment by afcowie** [comment]
@hvr This is extraordinary debugging work by @mpickering. If you look at the Haskell source for [HelloMain.hs](https://gist.github.com/mpickering/d335ac940a97a574891c947578636fa1#file-hellomain-hs) in Matt's Gist he worked down to a _very_ minimal example. Do you think you might have any insight about what's happening here? Amazing bug, really.

**Comment by hvr** [comment]
@mpicker @afcowie  I tried to reproduce this by cloning the gist (fixing up the .cabal file) and `cabal run`-ing with a couple GHC versions (on Ubuntu 18.04, with GHCs from my PPA) but wasn't able to reproduce yet; what environment did the segfaults occur on?

**Comment by mpickering** [comment]
@hvr I couldn't reproduce it with a normal GHC executable, only when using `ghcide`, which is why I didn't raise an upstream bug report. I tried a few things like normal GHC, ghci, ghci with `-fobject-code` and so on but couldn't get the segfault to trigger. 

Btw, I never said this in the issue but the segfault only happens if the string is at least 16 bytes long. Strings under 16 bytes seem to work.

**Comment by hvr** [comment]
@mpickering well, I tried with ghcide too and still wasn't able to reproduce:

```
$ cabal build -w ghc-8.8.3
Resolving dependencies...
Build profile: -w ghc-8.8.3 -O1
In order, the following will be built (use -v for more details):
 - hello-0.1.1 (exe:hello) (first run)
Configuring executable 'hello' for hello-0.1.1..
Preprocessing executable 'hello' for hello-0.1.1..
Building executable 'hello' for hello-0.1.1..
[1 of 2] Compiling Main             ( src/HelloMain.hs, /home/hvr/Haskell/bugs/text-short-mpickering/dist-newstyle/build/x86_64-linux/ghc-8.8.3/hello-0.1.1/x/hello/build/hello/hello-tmp/Main.o )
[2 of 2] Compiling Paths_hello      ( /home/hvr/Haskell/bugs/text-short-mpickering/dist-newstyle/build/x86_64-linux/ghc-8.8.3/hello-0.1.1/x/hello/build/hello/autogen/Paths_hello.hs, /home/hvr/Haskell/bugs/text-short-mpickering/dist-newstyle/build/x86_64-linux/ghc-8.8.3/hello-0.1.1/x/hello/build/hello/hello-tmp/Paths_hello.o )
Linking /home/hvr/Haskell/bugs/text-short-mpickering/dist-newstyle/build/x86_64-linux/ghc-8.8.3/hello-0.1.1/x/hello/build/hello/hello ...
```

```
$ PATH=/opt/ghc/8.8.3/bin:$PATH ghcide src/HelloMain.hs 
ghcide version: 0.1.0 (GHC: 8.8.3) (PATH: /stuff/dot-cabal/store/ghc-8.8.3/ghcide-0.1.0-e016e4e555362fde9aa6e177d286bd58be2d80403925ffef6459d40d32cb0a0e/bin/ghcide)
Ghcide setup tester in /home/hvr/Haskell/bugs/text-short-mpickering.
Report bugs at https://github.com/digital-asset/ghcide/issues…

Step 1/6: Finding files to test in /home/hvr/Haskell/bugs/text-short-mpickering
Found 1 files

Step 2/6: Looking for hie.yaml files that control setup
Found 1 cradle

Step 3/6, Cradle 1/1: Loading /home/hvr/Haskell/bugs/text-short-mpickering/hie.yaml

Step 4/6, Cradle 1/1: Loading GHC Session
> Resolving dependencies...
> Build profile: -w ghc-8.8.3 -O1
> In order, the following will be built (use -v for more details):
>  - hello-0.1.1 (exe:hello) (configuration changed)
> Configuring executable 'hello' for hello-0.1.1..
> Preprocessing executable 'hello' for hello-0.1.1..

Step 5/6: Initializing the IDE

Step 6/6: Type checking the files

Completed (1 file worked, 0 files failed)
```

**Comment by afcowie** [comment]
Bummer. Thanks for trying @hvr. I just tried with Matt's minimal example and it still segfaults the same way.

**Comment by mpickering** [comment]
@afcowie @hvr I am using NixOS, what about your Andrew? Perhaps I should try with a different C compiler?

**Comment by mpickering** [comment]
I just tried the example again with 8.8.1 and it worked, are you building a recent commit from master @afcowie ?

**Comment by afcowie** [comment]
@mpickering Fedora 31 Linux on an Intel i9; GHC 8.8.3 as specified by the `lts-15.5` Stackage snapshot (and brought in by **stack**) which brings in **text-short** 0.1.3. **ghcide** was master as at 27 March.

**Comment by domenkozar** [comment]
@afcowie if you're using ghcide-nix, try uncommenting https://github.com/cachix/ghcide-nix/blob/master/nix/default.nix#L19 and check if that fixes it.

**Comment by afcowie** [comment]
I have had some odd behaviours with `-dynamic` in the past - I had to stop [trying to] use that. Perhaps something built that way is lurking in my caches. I will blow away my _~/.stack- and rebuild fresh.

**Comment by afcowie** [comment]
> I will blow away my _~/.stack_ and rebuild fresh.

So I did a full from scratch re-installation of all the Haskell infrastructure, and unfortunately the problem endures.

**Comment by jneira** [comment]
Maybe this one is fixed by #836? @istathar could you give a try, please?

**Comment by istathar** [comment]
@jneira alas. I just tried it with **ghcide** from 'master' and I'm afraid to report that when I remove the `#ifdef` workaround it once again:

```
> Configuring GHCi with the following packages: publish
> /home/andrew/src/aesiniath/publish/.stack-work/install/x86_64-linux-tinfo6/bdcec458283ae25b661c440111c3bb5c26aa34a6c34f4902d47dcef25a0df0b9/8.8.4/pkgdb:/home/andrew/.stack/snapshots/x86_64-linux-tinfo6/bdcec458283ae25b661c440111c3bb5c26aa34a6c34f4902d47dcef25a0df0b9/8.8.4/pkgdb:/home/andrew/.stack/programs/x86_64-linux/ghc-tinfo6-8.8.4/lib/ghc-8.8.4/package.conf.d
[INFO] Using interface files cache dir: /home/andrew/.cache/ghcide/main-f9d60a41ddf9df52f88d70bf88584bd1798823c3
[INFO] Making new HscEnv[main]
Segmentation fault (core dumped)
```

**Comment by DylanSp** [comment]
I'm also getting this error with a Stack project on Linux (no Nix), initialized with `stack new todomvc-backend servant`; the problematic line is `$(deriveJSON defaultOptions ''User)`. My comment [here](https://github.com/haskell/haskell-language-server/issues/376#issuecomment-723367823) has the logs from running `haskell-language-server` against my code.
